File: advent-of-code/src/main/java/com/willmolloy/adventofcode/common/grid/Direction.java
package com.willmolloy.adventofcode.common.grid;

import java.util.List;

/**
 * 2D direction.
 *
 * @author <a href=https://willmolloy.com>Will Molloy</a>
 */
public enum Direction {
  UP,
  DOWN,
  LEFT,
  RIGHT,
  LEFT_UP,
  RIGHT_UP,
  LEFT_DOWN,
  RIGHT_DOWN;

  public static List<Direction> straight() {
    return List.of(UP, DOWN, LEFT, RIGHT);
  }

  public static List<Direction> straightAndDiagonal() {
    return List.of(Direction.values());
  }
}

File: advent-of-code/src/main/kotlin/com/willmolloy/adventofcode/common/grid/Direction.kt
package com.willmolloy.adventofcode.common.grid

/**
 * 2D direction.
 *
 * @author <a href=https://willmolloy.com>Will Molloy</a>
 */
enum class Direction {
  UP,
  DOWN,
  LEFT,
  RIGHT,
  LEFT_UP,
  RIGHT_UP,
  LEFT_DOWN,
  RIGHT_DOWN;

  companion object {
    @JvmStatic fun straight(): List<Direction> = listOf(UP, DOWN, LEFT, RIGHT)

    @JvmStatic fun straightAndDiagonal(): List<Direction> = entries
  }
}
